### PR TITLE
MFIter: Make Finalize Public

### PR DESCRIPTION
**Pull request by ax3l**
## Summary

Follow-up to #2983

## Checklist

The proposed changes:
- [ ] fix a bug or incorrect behavior in AMReX
- [ ] add new capabilities to AMReX
- [ ] changes answers in the test suite to more than roundoff level
- [ ] are likely to significantly affect the results of downstream AMReX users
- [ ] include documentation in the code and/or rst files, if appropriate

### CODE REVIEW

**Comment by ax3l** [comment]
Should we do something to `nextDynamicIndex` and `currentIndex` in `MFIter::Finalize`?

That would prevent that `operator++` (and other public member functions) create an invalid state on an already finalized `MFIter`.

**Comment by WeiqunZhang** [comment]
I guess we have to trust pyamrex users.

**Review by WeiqunZhang — APPROVED** [review]

**Comment by ax3l** [comment]
> I guess we have to trust pyamrex users.

They should not be able to call this, we just need it when implementing our [wrapper iterator](https://github.com/AMReX-Codes/pyamrex/blob/82d6260d85d70ecc8f3de679b8ed1fc9d8f53758/src/Base/MultiFab.cpp#L58-L80) while we are still in C++ :D
But can we trust other C++ users not to call it? :D

Jokes aside, should I just set both to the value of `endIndex` to add some minor safety?

I would like to be at least able to call `isValid()` on it:
https://github.com/AMReX-Codes/pyamrex/blob/82d6260d85d70ecc8f3de679b8ed1fc9d8f53758/src/Base/MultiFab.cpp#L44

File: Src/Base/AMReX_MFIter.cpp [hunk 1]
```diff
@@ -215,9 +215,13 @@ MFIter::~MFIter ()
 void
 MFIter::Finalize ()
 {
+    // avoid double finalize
     if (finalized) return;
     finalized = true;
 
+    // mark as invalid
+    currentIndex = endIndex;
```
Comment (WeiqunZhang) [inline Src/Base/AMReX_MFIter.cpp:223]: It won't hurt, although it's already invalid in normal usage, otherwise the MFIter loop won't stop.

Comment (ax3l) [reply Src/Base/AMReX_MFIter.cpp:223]: At least on our side, that's also the only location we would call this:
https://github.com/AMReX-Codes/pyamrex/blob/82d6260d85d70ecc8f3de679b8ed1fc9d8f53758/src/Base/MultiFab.cpp#L72-L75

I was thinking to add some safety if someone in C++ calls `mfi.Finalize()` mid loop - but why would they... maybe to abort an MFIter loop in a clean way...

Comment (WeiqunZhang) [reply Src/Base/AMReX_MFIter.cpp:223]: They can use `goto`.   That might be the cleanest way.  :smile:

**Pull request merged**